### ISSUE TITLE
[jvm-packages] Maven build issues during testing phase of xgboost4j-spark

### ISSUE THREAD

**Issue by qqilihq** [dmlc/xgboost#1553]
I'm having an issue with the Maven build of the `xgboost4j-spark` package. On two different machines the build is hanging during the test phase. The last log output is as follows:

```
16/09/07 15:19:49 INFO MemoryStore: Block rdd_1_0 stored as values in memory (estimated size 64.0 B, free 7.0 KB)
16/09/07 15:19:49 INFO BlockManagerInfo: Added rdd_1_0 in memory on localhost:61244 (size: 64.0 B, free: 1140.4 MB)
16/09/07 15:19:49 INFO Executor: Finished task 0.0 in stage 0.0 (TID 0). 1529 bytes result sent to driver
16/09/07 15:19:49 INFO TaskSetManager: Finished task 0.0 in stage 0.0 (TID 0) in 446 ms on localhost (3/4)
[4] train-error:0.000614
[15:19:49] src/tree/updater_prune.cc:74: tree pruning end, 1 roots, 8 extra nodes, 0 pruned nodes, max_depth=3
[3] train-error:0.000000
[15:19:49] src/tree/updater_prune.cc:74: tree pruning end, 1 roots, 8 extra nodes, 0 pruned nodes, max_depth=4
16/09/07 15:19:49 INFO MemoryStore: Block rdd_1_2 stored as values in memory (estimated size 64.0 B, free 7.1 KB)
16/09/07 15:19:49 INFO BlockManagerInfo: Added rdd_1_2 in memory on localhost:61244 (size: 64.0 B, free: 1140.4 MB)
16/09/07 15:19:49 INFO Executor: Finished task 2.0 in stage 0.0 (TID 2). 1529 bytes result sent to driver
16/09/07 15:19:49 INFO TaskSetManager: Finished task 2.0 in stage 0.0 (TID 2) in 444 ms on localhost (4/4)
16/09/07 15:19:49 INFO TaskSchedulerImpl: Removed TaskSet 0.0, whose tasks have all completed, from pool 
16/09/07 15:19:49 INFO DAGScheduler: ResultStage 0 (foreachPartition at XGBoost.scala:152) finished in 0,484 s
16/09/07 15:19:49 INFO DAGScheduler: Job 0 finished: foreachPartition at XGBoost.scala:152, took 0,503601 s
```

Then nothing happens any more. I left the terminal window open during the night, and got the following log lines this morning:

```
16/09/08 10:31:35 WARN HeartbeatReceiver: Removing executor driver with no recent heartbeats: 8384231 ms exceeds timeout 120000 ms
16/09/08 10:31:36 ERROR TaskSchedulerImpl: Lost executor driver on localhost: Executor heartbeat timed out after 8384231 ms
16/09/08 10:31:36 INFO DAGScheduler: Executor lost: driver (epoch 0)
16/09/08 10:31:36 INFO BlockManagerMasterEndpoint: Trying to remove executor driver from BlockManagerMaster.
16/09/08 10:31:36 INFO BlockManagerMasterEndpoint: Removing block manager BlockManagerId(driver, localhost, 61244)
16/09/08 10:31:36 WARN SparkContext: Killing executors is only supported in coarse-grained mode
16/09/08 10:31:36 INFO BlockManagerMaster: Removed driver successfully in removeExecutor
16/09/08 10:31:36 INFO DAGScheduler: Host added was in lost list earlier: localhost
16/09/08 10:31:36 INFO Executor: Told to re-register on heartbeat
16/09/08 10:31:36 INFO BlockManager: BlockManager re-registering with master
16/09/08 10:31:36 INFO BlockManagerMaster: Trying to register BlockManager
16/09/08 10:31:36 INFO BlockManagerMasterEndpoint: Registering block manager localhost:61244 with 1140.4 MB RAM, BlockManagerId(driver, localhost, 61244)
16/09/08 10:31:36 INFO BlockManagerMaster: Registered BlockManager
16/09/08 10:31:36 INFO BlockManager: Reporting 6 blocks to the master.
16/09/08 10:31:36 INFO BlockManagerInfo: Added rdd_1_3 in memory on localhost:61244 (size: 64.0 B, free: 1140.4 MB)
16/09/08 10:31:36 INFO BlockManagerInfo: Added broadcast_0_piece0 in memory on localhost:61244 (size: 2.3 KB, free: 1140.4 MB)
16/09/08 10:31:36 INFO BlockManagerInfo: Added rdd_1_0 in memory on localhost:61244 (size: 64.0 B, free: 1140.4 MB)
16/09/08 10:31:36 INFO BlockManagerInfo: Added rdd_1_1 in memory on localhost:61244 (size: 64.0 B, free: 1140.4 MB)
16/09/08 10:31:36 INFO BlockManagerInfo: Added rdd_1_2 in memory on localhost:61244 (size: 64.0 B, free: 1140.4 MB)
16/09/08 10:31:46 INFO Executor: Told to re-register on heartbeat
16/09/08 10:31:46 INFO BlockManager: BlockManager re-registering with master
16/09/08 10:31:46 INFO BlockManagerMaster: Trying to register BlockManager
16/09/08 10:31:46 INFO BlockManagerMaster: Registered BlockManager
16/09/08 10:31:46 INFO BlockManager: Reporting 6 blocks to the master.
16/09/08 10:31:46 INFO BlockManagerInfo: Added rdd_1_3 in memory on localhost:61244 (size: 64.0 B, free: 1140.4 MB)
16/09/08 10:31:46 INFO BlockManagerInfo: Added broadcast_0_piece0 in memory on localhost:61244 (size: 2.3 KB, free: 1140.4 MB)
16/09/08 10:31:46 INFO BlockManagerInfo: Added rdd_1_0 in memory on localhost:61244 (size: 64.0 B, free: 1140.4 MB)
16/09/08 10:31:46 INFO BlockManagerInfo: Added rdd_1_1 in memory on localhost:61244 (size: 64.0 B, free: 1140.4 MB)
16/09/08 10:31:46 INFO BlockManagerInfo: Added rdd_1_2 in memory on localhost:61244 (size: 64.0 B, free: 1140.4 MB)
16/09/08 10:31:56 INFO Executor: Told to re-register on heartbeat
16/09/08 10:31:56 INFO BlockManager: BlockManager re-registering with master
16/09/08 10:31:56 INFO BlockManagerMaster: Trying to register BlockManager
16/09/08 10:31:56 INFO BlockManagerMaster: Registered BlockManager
16/09/08 10:31:56 INFO BlockManager: Reporting 6 blocks to the master.
16/09/08 10:31:56 INFO BlockManagerInfo: Added rdd_1_3 in memory on localhost:61244 (size: 64.0 B, free: 1140.4 MB)
16/09/08 10:31:56 INFO BlockManagerInfo: Added broadcast_0_piece0 in memory on localhost:61244 (size: 2.3 KB, free: 1140.4 MB)
16/09/08 10:31:56 INFO BlockManagerInfo: Added rdd_1_0 in memory on localhost:61244 (size: 64.0 B, free: 1140.4 MB)
16/09/08 10:31:56 INFO BlockManagerInfo: Added rdd_1_1 in memory on localhost:61244 (size: 64.0 B, free: 1140.4 MB)
16/09/08 10:31:56 INFO BlockManagerInfo: Added rdd_1_2 in memory on localhost:61244 (size: 64.0 B, free: 1140.4 MB)
^C16/09/08 10:32:04 INFO RabitTracker$TrackerProcessLogger: Traceback (most recent call last):
16/09/08 10:32:04 INFO RabitTracker$TrackerProcessLogger:   File "/var/folders/55/tcpxvd8d4jd3lhr1rzyt1h840000gn/T/tracker7139232108892530423.py", line 475, in <module>
16/09/08 10:32:04 INFO RabitTracker$TrackerProcessLogger:     main()
16/09/08 10:32:04 INFO RabitTracker$TrackerProcessLogger:   File "/var/folders/55/tcpxvd8d4jd3lhr1rzyt1h840000gn/T/tracker7139232108892530423.py", line 470, in main
16/09/08 10:32:04 INFO RabitTracker$TrackerProcessLogger:     start_rabit_tracker(args)
16/09/08 10:32:04 INFO RabitTracker$TrackerProcessLogger:   File "/var/folders/55/tcpxvd8d4jd3lhr1rzyt1h840000gn/T/tracker7139232108892530423.py", line 441, in start_rabit_tracker
16/09/08 10:32:04 INFO RabitTracker$TrackerProcessLogger:     rabit.join()
16/09/08 10:32:04 INFO RabitTracker$TrackerProcessLogger:   File "/var/folders/55/tcpxvd8d4jd3lhr1rzyt1h840000gn/T/tracker7139232108892530423.py", line 331, in join
16/09/08 10:32:04 INFO RabitTracker$TrackerProcessLogger:     self.thread.join(100)
16/09/08 10:32:04 INFO RabitTracker$TrackerProcessLogger:   File "/usr/local/Cellar/python/2.7.12/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 951, in join
16/09/08 10:32:04 INFO RabitTracker$TrackerProcessLogger:     self.__block.wait(delay)
16/09/08 10:32:04 INFO RabitTracker$TrackerProcessLogger:   File "/usr/local/Cellar/python/2.7.12/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 359, in wait
16/09/08 10:32:04 INFO RabitTracker$TrackerProcessLogger:     _sleep(delay)
16/09/08 10:32:04 INFO RabitTracker$TrackerProcessLogger: KeyboardInterrupt
16/09/08 10:32:04 INFO SparkContext: Invoking stop() from shutdown hook
16/09/08 10:32:04 INFO RabitTracker$TrackerProcessLogger: Tracker Process ends with exit code 1
16/09/08 10:32:04 INFO RabitTracker: Tracker Process ends with exit code 1
16/09/08 10:32:04 INFO XGBoostSpark: Rabit returns with exit code 1
16/09/08 10:32:04 INFO SparkUI: Stopped Spark web UI at http://localhost:4040
16/09/08 10:32:04 INFO MapOutputTrackerMasterEndpoint: MapOutputTrackerMasterEndpoint stopped!
16/09/08 10:32:04 INFO MemoryStore: MemoryStore cleared
16/09/08 10:32:04 INFO BlockManager: BlockManager stopped
16/09/08 10:32:04 INFO BlockManagerMaster: BlockManagerMaster stopped
16/09/08 10:32:04 INFO OutputCommitCoordinator$OutputCommitCoordinatorEndpoint: OutputCommitCoordinator stopped!
16/09/08 10:32:04 INFO SparkContext: Successfully stopped SparkContext
16/09/08 10:32:04 INFO RemoteActorRefProvider$RemotingTerminator: Shutting down remote daemon.
16/09/08 10:32:04 INFO RemoteActorRefProvider$RemotingTerminator: Remote daemon shut down; proceeding with flushing remote transports.
16/09/08 10:32:04 INFO ShutdownHookManager: Shutdown hook called
16/09/08 10:32:04 INFO RemoteActorRefProvider$RemotingTerminator: Remoting shut down.
16/09/08 10:32:04 INFO ShutdownHookManager: Deleting directory /private/var/folders/55/tcpxvd8d4jd3lhr1rzyt1h840000gn/T/spark-bd9b9222-ca21-46ce-a5bf-f22160c6f195
- training with external memory cache *** FAILED ***
  ml.dmlc.xgboost4j.java.XGBoostError: XGBoostModel training failed
  at ml.dmlc.xgboost4j.scala.spark.XGBoost$.train(XGBoost.scala:169)
  at ml.dmlc.xgboost4j.scala.spark.XGBoostSuite$$anonfun$4.apply$mcV$sp(XGBoostSuite.scala:155)
  at ml.dmlc.xgboost4j.scala.spark.XGBoostSuite$$anonfun$4.apply(XGBoostSuite.scala:143)
  at ml.dmlc.xgboost4j.scala.spark.XGBoostSuite$$anonfun$4.apply(XGBoostSuite.scala:143)
  at org.scalatest.Transformer$$anonfun$apply$1.apply$mcV$sp(Transformer.scala:22)
  at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
  at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
  at org.scalatest.Transformer.apply(Transformer.scala:22)
  at org.scalatest.Transformer.apply(Transformer.scala:20)
  at org.scalatest.FunSuiteLike$$anon$1.apply(FunSuiteLike.scala:166)
  ...
[4] train-error:0.000000
```

Both machines are running OS X 10.11.6 with Java 1.8.0_66.

Spark is not relevant to me currently, so I just ignore the issue however I thought it would be worth reporting the problem.

**Comment by xydrolase** [comment]
Perhaps some connection issues to the Rabit tracker. I ran into the same issues when running `xgboost4j-spark` on a Spark cluster, using the default port number for the Rabit tracker. The issue seemed to resolve when I modified the Python code of the tracker to use a random port from the range 32768 - 61000. 

The fact that the Rabit tracker is in another project (dmlc), and implemented in Python makes it hard to debug. I'm working on a pure Scala implementation of the Rabit tracker pending tests. 

Will send a pull request once the code proves to be stable.
